amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/impl/PersistentQueue.java

```diff
@@ -178,12 +194,68 @@ private List<AmqpQueueProperties> getQueueProperties(Map<String, AmqpMessageRout
     private void topicNameValidate() {
         String[] nameArr = this.indexTopic.getName().split("/");
         checkArgument(nameArr[nameArr.length - 1].equals(TOPIC_PREFIX + queueName),
                 "The queue topic name does not conform to the rules(%s%s).",
                 TOPIC_PREFIX, "exchangeName");
     }
 
+    private void scheduleExchangeClearTask() {
+        this.indexTopic.getBrokerService().getPulsar().getExecutor()
+                .schedule(this::exchangeClearCheckpointTask, exchangeClearTaskInterval, TimeUnit.MILLISECONDS);
+    }
+
+    private synchronized void exchangeClearCheckpointTask() {
+        this.checkpointLac = null;
+        Collection<CompletableFuture<Void>> futures = new ArrayList<>();
+        for (AmqpMessageRouter router : routers.values()) {
+            CompletableFuture<Void> future = new CompletableFuture<>();
+            ((PersistentTopic) router.getExchange().getTopic()).getManagedLedger()
+                    .asyncOpenCursor("__amqp_replicator__" + router.getExchange().getName(),
+                            new AsyncCallbacks.OpenCursorCallback() {
+                @Override
+                public void openCursorComplete(ManagedCursor cursor, Object ctx) {
+                    PositionImpl pos = (PositionImpl) cursor.getMarkDeletedPosition();
+                    checkpointExchangeRoutePos.put(router.getExchange().getName(), pos);
+                    future.complete(null);
+                }
+
+                @Override
+                public void openCursorFailed(ManagedLedgerException exception, Object ctx) {
+                    log.warn("Failed to get exchange route position for queue {}.", queueName, exception);
+                    future.completeExceptionally(exception);
+                }
+            }, null);
+            futures.add(future);
+        }
+        FutureUtil.waitForAll(futures).whenComplete((__, t) -> {
+            if (t != null) {
+                scheduleExchangeClearTask();
+                return;
+            }
+            this.checkpointLac = (PositionImpl) indexTopic.getLastPosition();
+            exchangeCleanup();
+        });
+    }
+
+    private void exchangeCleanup() {
+        PositionImpl indexMarkDeletePos = (PositionImpl) indexTopic.getSubscription(DEFAULT_SUBSCRIPTION)
+                .getCursor().getMarkDeletedPosition();
+        Collection<CompletableFuture<Void>> futures = new ArrayList<>();
+        if (indexMarkDeletePos.compareTo(this.checkpointLac) >= 0) {
+            for (AmqpMessageRouter router : routers.values()) {
+                Position position = checkpointExchangeRoutePos.get(router.getExchange().getName());
+                if (position != null) {
+                    futures.add(router.getExchange().markDeleteAsync(
+                            queueName, position.getLedgerId(), position.getEntryId()));
+                }
+            }
+            FutureUtil.waitForAll(futures).thenRun(this::scheduleExchangeClearTask);
```

Comment:
When (router.getExchange().markDeleteAsync) throws an exception, the task will stop. Replace (thenRun) with (whenComplete)

Reply:
Good catch.

Fixed, PTAL